python/ql/src/experimental/semmle/python/frameworks/Xml.qll

```diff
@@ -0,0 +1,301 @@
+   *
+   * * `this` would be `xml.etree.ElementTree.fromstring(xml_content, parser=parser)`.
+   * * `getAnInput()`'s result would be `xml_content`.
+   * * `vulnerable(kind)`'s `kind` would be `XXE`.
+   */
+  private class XMLEtreeParsing extends DataFlow::CallCfgNode, XML::XMLParsing::Range {
+    XMLEtreeParsing() {
+      this =
+        API::moduleImport("xml")
+            .getMember("etree")
+            .getMember("ElementTree")
+            .getMember(["fromstring", "fromstringlist", "XML", "parse"])
+            .getACall()
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate vulnerable(string kind) {
+      exists(XML::XMLParser xmlParser |
+        xmlParser = this.getArgByName("parser").getALocalSource() and xmlParser.vulnerable(kind)
+      )
+    }
+  }
+
+  /** Gets a reference to a `parser` that has been set a `feature`. */
+  private DataFlow::Node trackSaxFeature(
+    DataFlow::TypeTracker t, DataFlow::CallCfgNode parser, API::Node feature
+  ) {
+    t.start() and
+    exists(DataFlow::MethodCallNode featureCall |
+      featureCall = parser.getAMethodCall("setFeature") and
+      featureCall.getArg(0).getALocalSource() = feature.getAUse() and
+      featureCall.getArg(1).getALocalSource() = DataFlow::exprNode(any(True t_)) and
+      result = featureCall.getObject()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 |
+      t = t2.smallstep(trackSaxFeature(t2, parser, feature), result)
+    )
+  }
+
+  /** Gets a reference to a `parser` that has been set a `feature`. */
+  DataFlow::Node trackSaxFeature(DataFlow::CallCfgNode parser, API::Node feature) {
+    result = trackSaxFeature(DataFlow::TypeTracker::end(), parser, feature)
+  }
+
+  /**
+   * Gets a call to `xml.sax.make_parser`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   * BadHandler = MainHandler()
+   * parser = xml.sax.make_parser()
+   * parser.setContentHandler(BadHandler)
+   * parser.setFeature(xml.sax.handler.feature_external_ges, False)
+   * parser.parse(StringIO(xml_content))
+   * parsed_xml = BadHandler._result
+   * ```
+   *
+   * * `this` would be `xml.sax.make_parser()`.
+   * * `getAnInput()`'s result would be `StringIO(xml_content)`.
+   * * `vulnerable(kind)`'s `kind` would be `Billion Laughs` and `Quadratic Blowup`.
+   */
+  private class XMLSaxParser extends DataFlow::CallCfgNode, XML::XMLParser::Range {
+    XMLSaxParser() {
+      this = API::moduleImport("xml").getMember("sax").getMember("make_parser").getACall()
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getAMethodCall("parse").getArg(0) }
+
+    override predicate vulnerable(string kind) {
+      exists(DataFlow::MethodCallNode parse, API::Node handler, API::Node feature |
+        handler = API::moduleImport("xml").getMember("sax").getMember("handler") and
+        parse.calls(trackSaxFeature(this, feature), "parse") and
+        parse.getArg(0) = this.getAnInput() // enough to avoid FPs?
```

Comment:
I'm not sure if this would be enough to avoid FPs when multiple calls and inputs collide in the same flow :/